Extend OutputStyle

peterrum opened the pull request:
A first draft. An open question is how to treat `Description` and if we need `DefaultStyle`.

The many changed loc are related to the fact that some switch-cases had to be replaced by if-else constructions.

Closes #9888 and #9903.

nfehn commented on source/base/parameter_handler.cc:
```diff
@@ -1329,285 +1328,269 @@ ParameterHandler::recursively_print_parameters(
 
   // this function should not be necessary for XML or JSON output...
-  Assert((style != XML) && (style != JSON), ExcInternalError());
+  Assert(((style & XML) != XML) && ((style & JSON) != JSON),
```
XML has to be written twice, JSON has to be written twice

bangerth replied:
You could use `count_set_bits(style & (XML | JSON | Text)) == 1` if you had a function that counts the number of bits set in an integer. This way you make sure that exactly one bit is set.

An implementation of such a function is here: https://www.geeksforgeeks.org/count-set-bits-in-an-integer/

bangerth replied:
Better choice is here: https://stackoverflow.com/questions/109023/how-to-count-the-number-of-set-bits-in-a-32-bit-integer

bangerth replied:
Or, with C++20: https://en.cppreference.com/w/cpp/numeric/popcount

nfehn commented on source/base/parameter_handler.cc:
```diff
@@ -1276,35 +1275,35 @@ ParameterHandler::print_parameters(std::ostream &    out,
 
   // for all of the other formats, print a preamble:
-  switch (style)
+  if ((style & (Text | Short)) == (Text | Short))
```
`Text | Short` has to be written twice

nfehn commented on source/base/parameter_handler.cc:
```diff
@@ -1738,7 +1717,7 @@ ParameterHandler::log_parameters_section(LogStream &out,
 
   // Sort parameters alphabetically, if needed.
-  if (sort_alphabetical)
+  if (!(style & Unsorted))
```
double negation

nfehn commented:
I have another general question: Where is the correctness check done that excludes invalid parameter combinations?

nfehn commented on include/deal.II/base/parameter_handler.h:
```diff
@@ -851,52 +851,59 @@ class ParameterHandler : public Subscriptor
+    LaTeX = 0x0020,
+
     /**
      * Write out declared parameters with description and possible values.
```
sure, but don't you see the difficulties in documentation? Since we don't have a struct design, which would more intuitively show to which part this parameter belongs to, we should document it. My point is that this parameter can be (mis-)understood as the boolean `Short` (or its opposite).

luca-heltai commented:
@peterrum , as a followup to @nfehn comments, could you please include a better documentation of the `Description` format, i.e., specifying that it is actually a format, and not an additional style (like documentation/alphabetical, etc.)?

I think these need to be addressed, and we can proceeed with merging this PR, also according to the votes in #9888:

- [ ] better documentation of the `Description` format
- [ ] adding some `AssertThrow` at the beginning of the print and log functions, that checks that only one output format is used at a given time
- [ ] Making a `DefaultStyle` that does NOT specify a format, but simply the current default stylistic options (i.e., alphabetical, documented)
- [ ] add `operator|` for `OutputStyle`

nfehn commented:
I have only two minor comments:

1.

> adding some AssertThrow at the beginning of the print and log functions, that checks that only one output format is used at a given time

It would be good to have these checks at only once place in the code. Why not in the ``operator|`` function? Otherwise, we would actually have to go back to discussion #9888 and rethink certain concepts.

2. Could we please merge #9862 first?

luca-heltai commented:
I think the interface of #9862 should change back to be compatible with 9.1. I think it is better to do this after we have fixed this one.

luca-heltai commented:
> Why not in the operator| function?

This is also a good idea, but I would keep the Assert in the code nonetheless.

peterrum commented:
@luca-heltai I'll tackle this PR this evening. I need to finish one or two things.

nfehn commented:
> I think the interface of #9862 should change back to be compatible with 9.1. I think it is better to do this after we have fixed this one.

Why? I did not change an interface in #9862. I relied on the interface of deal.II at that time. This pull request here and the related discussions in #9888 emerged later.

luca-heltai commented:
>  Why? I did not change an interface in #9862. I relied on the interface of deal.II at that time. This pull request here and the related discussions in #9888 emerged later.

which we are going to revert here.

nfehn commented:
@luca-heltai 

> but I would keep the Assert in the code nonetheless.

Why? Because more asserts help more? Don't you think #9888 would have been the place to discuss this and communicate the downsides of your favorite version? So, with every new function you want to add new asserts to the code, where it is unclear which part of the code is actually responsible for that? There were suggestions in #9888 that do not need this kind of asserts here and there in the code.

tamiko commented:
Dear all, I would really like to read through the `ParameterHandler` changes without such an aggressive tone everywhere :disappointed:

In particular, we are discussing indredibly minor changes here. So the core principles should be (a) not to break our interfaces from previous releases, and (b) provide a reasonably extensible and intuitive interface.

That said I think it makes a lot of sense to (a) avoid bool, (b) retain and mildly abuse `OutputStyle` as well.

I have only one suggestion, though. I see that we need a number of rather finicky bit comparisons. I suggest to give them a name:
```
// extract information from supplied `OutputStyle`
const bool sort_alphabetically = ...;
```

peterrum commented:
@luca-heltai May I ask you for a second round of review. In particular, I have updated the documentation and added some asserts at the front of the functions. I think it is fine not to have asserts in `operator|` since during construction of the final configuration it might be valid that a non-valid configuration is passed.

luca-heltai commented on source/base/parameter_handler.cc:
```diff
@@ -1219,18 +1219,28 @@ ParameterHandler::set(const std::string &entry_string, const bool new_value)
   AssertThrow(out, ExcIO());
 
+  Assert(
+    (((style & XML) > 0) + ((style & JSON) > 0) + ((style & PRM) > 0) +
+     ((style & Description) > 0) + ((style & LaTeX) > 0)),
```
Isn't a `==1` missing?
```suggestion
    (((style & XML) > 0) + ((style & JSON) > 0) + ((style & PRM) > 0) +
     ((style & Description) > 0) + ((style & LaTeX) > 0)) == 1,
```

bangerth replied:
Funny question that I'm not sure about: Say, `OutputStyle` uses `short unsigned int` as base type. The `style & XML` is promoted to `int`. Is the result guaranteed to be `>0` is one bit is set, or could this result in the sign bit being set and the result is negative? I think what you *want* to test is that at least one bit is set, so maybe using `!= 0` is the better check over `>0`?

luca-heltai commented on source/base/parameter_handler.cc:
```diff
@@ -1219,18 +1219,28 @@ ParameterHandler::set(const std::string &entry_string, const bool new_value)
 {
   AssertThrow(out, ExcIO());
 
+  Assert(
```
This and the others should be `AssertThrow`. In principle, the output style could be decided at run time, by a parsed string. We want to make sure we assert also in Release mode.

luca-heltai commented on source/base/parameter_handler.cc:
```diff
@@ -1329,285 +1339,270 @@ ParameterHandler::recursively_print_parameters(
 
-      default:
-        Assert(false, ExcNotImplemented());
+            // if there is a documenting string, print it as well
```
Can you surround this in a `if(!(style & Short))`? This way we'll have also the `ShortDescription` implemented.

Also add this to the possible combinations, and fix the table at the beginning.

luca-heltai commented:
Doxygen is complaining for some reasons, but I could not find exactly where:

~~~
$ cat doxygen.log && ! [ -s doxygen.log ]
/home/travis/build/dealii/dealii/include/deal.II/base/parameter_handler.h:954: warning: reached end of file while inside a code block!
The command that should end the block seems to be missing!
The command "cat doxygen.log && ! [ -s doxygen.log ]" exited with 1.
~~~

luca-heltai commented:
I think it is referring to the constructs `<code>PRM<\code> instead of <code>Text<\code>.`

Can you replace those by their mardkdown counterpart (using backticks)?

luca-heltai commented:
/rebuild

nfehn commented on source/base/parameter_handler.cc:
```diff
@@ -1219,18 +1219,28 @@ ParameterHandler::set(const std::string &entry_string, const bool new_value)
+  Assert(((style & LaTeX) && (style & Short)) == false,
+         ExcMessage("It is not possible to request short LaTex."));
+  Assert(((style & Description) && (style & Short)) == false,
+         ExcMessage("It is not possible to request short Description."));
```
Personally, I would prefer to have all these asserts in one private member function that ensures consistency of the output style, as a means to satisfy our design goals from #9888 and to highlight that this check is related to the data structure OutputStyle and not to a specific function? If I understood it correctly, more functions might need these checks, so a copy-paste of this code is forseeable otherwise.

luca-heltai replied:
That’s actually a very nice suggestion. @peterrum, could you wrap up the checks in a function returning a Boolean, and assert on the returned value?

peterrum replied:
I went with @bangerth' suggestion with the free function in an anonymous namespace.

peterrum commented:
@luca-heltai  That should work. Locally it looks like this:

![OutputStyle](https://user-images.githubusercontent.com/15707929/80002955-a11f6d00-84c0-11ea-91a4-f7deb2230ea0.png)

I have applied all other suggestions.

luca-heltai commented:
@peterrum, would you mind implementing the other suggestion by @nfehn, i.e., wrap up the check of the consistency in a private function that returns a boolean, and assert on the return value instead of explicitly writing the  checks in the function? This will be useful also elsewhere.

bangerth commented:
Or a free function in an anonymous namespace in the .cc file if it doesn't require access to member variables.

bangerth commented on include/deal.II/base/parameter_handler.h:
```diff
@@ -846,57 +846,113 @@ class ParameterHandler : public Subscriptor
+     * Default stylistic style: print documentation and sort all parameters
+     * alphabetically.
+     */
+    DefaultStyle = 0x0000,
```
I regret that we hadn't settled on `use_underscore` instead of `CamelCase` for member variables and enum values back when this class was written :-(

But it's too late now. It makes more sense to keep the members of this class consistent with each other than to have some kind of mixture.

peterrum replied:
@bangerth We can still deprecate all old names. In that case, the question would be how to call `Short` since `short` is already taken.

luca-heltai replied:
Let's discuss this in another PR...

bangerth replied:
Yes. I don't think there is much of an upside to making such a change. It is what it is.

nfehn commented:
Did we forget to update the Assert in the second half of ``ParameterAcceptor::initialize()`` (if-branch ``if (!output_filename.empty()){...}`` in this PR?

https://www.dealii.org/developer/doxygen/deal.II/parameter__acceptor_8cc_source.html
